### DISCUSSION TITLE
type annotation for TypeDecorator

### DISCUSSION THREAD

**Discussion by finswimmer** [sqlalchemy/sqlalchemy#8660] [Usage Questions]
Hey,

I'm not sure if  this question is more a typing question in general or if it sqlalchemy specific.

I've created a custom TypeDecorator which implements TimeStamps. Doing this the column accepts a timestamp as string or `datetime`:

```py
from datetime import datetime, timezone
from pathlib import Path
from typing import TYPE_CHECKING, Any, Optional, Type, Union

import sqlalchemy.orm
from sqlalchemy import Column, Integer, create_engine, types
from sqlalchemy.ext.declarative import declarative_base

if TYPE_CHECKING:
    from datetime import tzinfo

Base = declarative_base()


def get_local_timezone() -> Optional["tzinfo"]:
    return datetime.utcnow().astimezone().tzinfo


class UtcTimeStamp(types.TypeDecorator):
    """Make sure datetimes going to db or coming from db are in utc."""

    impl = types.TIMESTAMP
    cache_ok = True

    @property
    def python_type(self) -> Type[datetime]:
        return datetime

    def process_bind_param(
        self, value: Optional[Union[datetime, str]], dialect: Any
    ) -> Optional[datetime]:
        """Convert python type to column type."""
        if not value:
            return None

        if isinstance(value, str):
            value = datetime.fromisoformat(value)

        if value.tzinfo is None:
            value = value.replace(tzinfo=get_local_timezone())

        return value.astimezone(timezone.utc)

    def process_result_value(
        self, value: Optional[datetime], dialect: Any
    ) -> Optional[datetime]:
        """Convert column type to python type."""
        if not value:
            return None

        if value.tzinfo is None:
            value = value.replace(tzinfo=timezone.utc)

        return value.astimezone(timezone.utc)


class ToDo(Base):
    __tablename__ = "todos"

    id: Optional[int] = Column(Integer, autoincrement=True, primary_key=True)
    start: datetime = Column(UtcTimeStamp(), nullable=False)


if Path("typehints.db").exists():
    Path("typehints.db").unlink()

engine = create_engine("sqlite:///typehints.db", echo="debug")

Base.metadata.create_all(engine)

todo = ToDo(start="2021-04-26 14:19:33.516423")
todo_2 = ToDo(start=datetime.now())

session = sqlalchemy.orm.Session(autocommit=False, autoflush=False, bind=engine)

session.add_all([todo, todo_2])
session.commit()
```

As you can see, I annotated the columns to have proper code completion in my IDE. `mypy` have legitimate concerns here:

```
error: Argument "start" to "ToDo" has incompatible type "str"; expected "datetime"  [arg-type]
```

I wonder if it is possible to type annotate `start` in a way, that it is saying it excepts `str` and `datetime` on setting, but always return `datetime`?

**Comment by CaselIT** [comment]
Hi,

I'm assuming you are using 1.4 with the mypy plugin.
I don't think it's possible unless you were to provide an init method

**Reply by finswimmer** [reply]
Thanks for your reply @CaselIT.

> I'm assuming you are using 1.4 with the mypy plugin.

You are assuming right. I've seen that 2.0 beta is released. So if something like this is possible there, I would be interested too.

> I don't think it's possible unless you were to provide an init method

I suspected it. An init method doesn't solve this completely I guess, because this value might also be set/changed later. So probably a `@property` / `@hybrid_property` is needed here as well.

**Reply by CaselIT** [reply]
> So if something like this is possible there, I would be interested too.

I don't think so, at least not that I know. You would need to have two different types one that's created by the orm that has datetime and one that's created by user code that could have datetime or string.

Maybe you could hack something together using generic but I'm not sure you can mix it with sqlalchemy declarative
